package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -699,27 +701,18 @@ def _filter_atoms(self, donors, acceptors):
     def _prepare(self):
         self.results.hbonds = [[], [], [], [], [], []]
 
-        # Set atom selections if they have not been provided
-        if self.acceptors_sel is None:
-            self.acceptors_sel = self.guess_acceptors()
-        if self.hydrogens_sel is None:
-            self.hydrogens_sel = self.guess_hydrogens()
-
         # Select atom groups
-        self._acceptors = self.u.select_atoms(self.acceptors_sel,
-                                              updating=self.update_selections)
+        if self.acceptors_sel is None:
+            self._acceptors = self.guess_acceptors()
+        else:
+            self._acceptors = self.u.select_atoms(self.acceptors_sel,
+                                                updating=self.update_selections)
         self._donors, self._hydrogens = self._get_dh_pairs()
 
     def _single_frame(self):
 
         box = self._ts.dimensions
 
-        # Update donor-hydrogen pairs if necessary
-        if self.update_selections:
-            self._donors, self._hydrogens = self._get_dh_pairs()
```

Comment:
I think this is still needed for `_single_frame` because `self._donors` will be updated every frame with the change of `self._hydrogens`. You need to separate the part that creates `self._hydrogens` from it

Reply:
@lunamorrow The reason you get different results is that in the old code, `self._donors` is updated every frame based on `self._hydrogens`. However, if you don't execute `_get_dh_pairs` every frame, `self._donors` will become a non-updating AtomGroup based on the first frame.